dali/test/python/test_dali_tf_dataset_pipelines.py

```diff
@@ -0,0 +1,271 @@
+            if "gpu" in device_str:
+                input_dataset = input_dataset.apply(tf.data.experimental.copy_to_device('/gpu:0'))
+
+        dataset_pipeline, shapes, dtypes = pipeline_desc
+
+        with tf.device(device_str):
+            dali_dataset = dali_tf.experimental.DALIDatasetWithInputs(
+                    input_datasets=input_dataset,
+                    input_names="input_placeholder",
+                    pipeline=dataset_pipeline,
+                    batch_size=dataset_pipeline.batch_size,
+                    output_shapes=shapes,
+                    output_dtypes=dtypes,
+                    num_threads=dataset_pipeline.num_threads,
+                    device_id=dataset_pipeline.device_id)
+        return dali_dataset
+    return to_dataset
+
+@nottest
+def external_source_tester(shape, dtype, source=None, external_source_device="cpu"):
+    def get_external_source_pipeline_getter(batch_size, num_threads, device, device_id=0,
+        shard_id=0, num_shards=1, def_for_dataset=False):
+
+        pipe = one_input_pipeline(def_for_dataset,
+                                  device,
+                                  source,
+                                  external_source_device,
+                                  batch_size=batch_size,
+                                  num_threads=num_threads,
+                                  device_id=device_id)
+
+        batch_shape = (batch_size,) + tuple(None for _ in shape)
+
+        return pipe, (batch_shape, batch_shape), (tf.dtypes.as_dtype(dtype), tf.int32)
+    return get_external_source_pipeline_getter
+
+
+@pipeline_def
+def many_input_pipeline(def_for_dataset, device, sources, input_names):
+    """ Pipeline accepting multiple inputs via external source
+
+    Parameters
+    ----------
+    def_for_dataset : bool
+         True if this pipeline will be converted to TF Dataset
+    device : str
+        device that the Dataset will be placed ("cpu" or "gpu")
+    sources : list of callables
+        callbacks for the external sources in baseline pipeline otherwise None
+    input_names : list of str
+        Names of inputs placeholder for TF
+    """
+    inputs = []
+    if def_for_dataset:
+        for input_name in input_names:
+            input = fn.external_source(name=input_name)
+            input = input if device == 'cpu' else input.gpu()
+            inputs.append(input)
+    else:
+        for source in sources:
+            input = fn.external_source(source=source, batch=False)
+            input = input if device == 'cpu' else input.gpu()
+            inputs.append(input)
+    processed = []
+    for input in inputs:
+        processed.append(fn.cast(input + 10, dtype=dali.types.INT32))
+    results = fn.pad(inputs + processed)
+    return tuple(results)
+
+
+# Test that uses multiple Generator dataset as inputs to DALI pipeline
+def external_source_converter_multiple(start_values, input_names):
```

Comment:
This same pattern is repeated at least 3 times with minimal changes. How do you feel about unifying to a common function with different options?

Reply:
I can see a pattern emerging a bit, but it's already a function creating a function. The TF dataset part is different, but now you can see it in one go. I don't want to obfuscate it more.

Reply:
I think that some redundancy is a fair price for simplicity in tests - the tests should make it easy to disable parts to pinpoint regressions, should any appear.